src/resource.rs

```diff
@@ -2,35 +2,313 @@
+            check_arg(a, p).map_err(SignatureError::TypeArgMismatch)?;
+        }
+        let (ins, outs, res) = match &self.signature_func {
+            SignatureFunc::FromYAML { .. } => {
+                // Sig should be computed solely from inputs + outputs + args.
+                todo!()
+            }
+            SignatureFunc::CustomFunc(bf) => bf.compute_signature(&self.name, args, &self.misc)?,
+        };
+        assert!(res.contains(&self.resource));
+        let mut sig = Signature::new_df(ins, outs);
+        sig.input_resources = resources_in.clone();
+        sig.output_resources = res.union(resources_in); // Pass input requirements through
+        Ok(sig)
+    }
+
+    /// Optional description of the ports in the signature.
+    pub fn signature_desc(&self, args: &[TypeArg]) -> SignatureDescription {
+        match &self.signature_func {
+            SignatureFunc::FromYAML { .. } => {
+                todo!()
+            }
+            SignatureFunc::CustomFunc(bf) => bf.describe_signature(&self.name, args, &self.misc),
+        }
+    }
+
+    pub(crate) fn should_serialize_signature(&self) -> bool {
+        match self.signature_func {
+            SignatureFunc::CustomFunc(_) => true,
+            SignatureFunc::FromYAML { .. } => false,
+        }
+    }
+
+    /// Fallibly returns a Hugr that may replace an instance of this OpDef
+    /// given a set of available resources that may be used in the Hugr.
+    pub fn try_lower(&self, args: &[TypeArg], available_resources: &ResourceSet) -> Option<Hugr> {
+        match &self.lower_func {
+            LowerFunc::None => None,
+            LowerFunc::FixedHugr(req_res, h) => {
+                if available_resources.is_superset(req_res) {
+                    Some(h.clone())
+                } else {
+                    None
+                }
+            }
+            LowerFunc::CustomFunc(f) => {
+                f.try_lower(&self.name, args, &self.misc, available_resources)
+            }
+        }
+    }
+}
 
-use crate::ops::{OpDef, OpaqueOp};
-use crate::types::custom::CustomType;
+/// A declaration of an opaque type.
+/// Note this does not provide any way to create instances
+/// - typically these are operations also provided by the Resource.
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+pub struct TypeDef {
+    /// The unique name of the type
+    pub name: SmolStr,
+    /// Declaration of type parameters. The TypeDef must be instantiated
+    /// with the same number of [`TypeArg`]'s to make an actual type.
+    ///
+    /// [`TypeArg`]: crate::types::type_param::TypeArg
+    pub args: Vec<TypeParam>,
+}
 
 /// A unique identifier for a resource.
 ///
 /// The actual [`Resource`] is stored externally.
 pub type ResourceId = SmolStr;
 
 /// A resource is a set of capabilities required to execute a graph.
-#[derive(Clone, Debug, Default, serde::Serialize, serde::Deserialize)]
+#[derive(Debug, Default, serde::Serialize, serde::Deserialize)]
```

Comment:
Yah, I am not sure about this. I do not *expect* to clone Resources; they'll be pretty big, and may contain third-party implementations of traits amongst others, so the intention is that the design does not require cloning. (This was also the motivation for `Arc<OpDef>`)

Reply:
ON second thoughts, yes I guess we should. Someone will want to build a Resource by cloning another and mutating the copy... (indeed, this is even cheaper than it was before the use of `Arc<OpDef>`!). I would like to avoid multiple Resources with the same name, but that will have to be enforced via the "registry" (effectively a `Map<String, Resource>`, here only used in `resolve_extension_ops` with the rest a TODO). And I think we probably should not be cloning Resources within the Hugr crate, but I guess that's not a reason not to derive Clone.

What would you think to a method `clone_and_rename(new_name: String) -> Self` instead/as well, tho?

Reply:
Yes, we will never clone a resource and re-use it with the same Hugr.

But in general having clone is more for some meta-tasks (e.g. testing). If
- there's not logical error in having multiple copies (as long as you don't use it on the same registry)
- they do not manage external resources (so no handles / destructor shenanigans)
- they are safe and cheap to move
there's no reason not to have `Clone`.

Reply:
Yes, I was thinking of testing and heading for the same conclusion. Done. (It's quite nice that the Arc means cloning a Resource doesn't try to clone the CustomSignatureFunc's :-) )